src/yubikey/piv/management.rs

```diff
@@ -142,45 +171,52 @@ impl super::Yubikey {
     /// Generate CSR for slot
     pub fn generate_csr(&mut self, slot: &SlotId, common_name: &str) -> Result<Vec<u8>> {
         let mut params = rcgen::CertificateParams::new(vec![]);
-        params.alg = match self.configured(slot)? {
-            PublicKeyInfo::EcP256(_) => &rcgen::PKCS_ECDSA_P256_SHA256,
-            PublicKeyInfo::EcP384(_) => &rcgen::PKCS_ECDSA_P384_SHA384,
+        let cert = self.configured(&slot).unwrap();
+        let pki = cert.subject_pki();
+        params.alg = match pki.algorithm.oid {
+            NISTP256_OID => &rcgen::PKCS_ECDSA_P256_SHA256,
+            SECG384_OID => &rcgen::PKCS_ECDSA_P384_SHA384,
             _ => return Err(Error::Unsupported),
         };
         params
             .distinguished_name
             .push(rcgen::DnType::CommonName, common_name.to_string());
 
         let csr_signer = CSRSigner::new(self.yk.serial().into(), *slot);
-        params.key_pair = Some(rcgen::KeyPair::from_remote(Box::new(csr_signer)).map_err(|e| Error::InternalYubiKeyError(format!("{}", e)))?);
+        params.key_pair = Some(
+            rcgen::KeyPair::from_remote(Box::new(csr_signer))
+                .map_err(|e| Error::InternalYubiKeyError(format!("{}", e)))?,
+        );
 
-        let csr = rcgen::Certificate::from_params(params).map_err(|e| Error::InternalYubiKeyError(format!("{}", e)))?;
-        let csr = csr.serialize_request_der().map_err(|e| Error::InternalYubiKeyError(format!("{}", e)))?;
+        let csr = rcgen::Certificate::from_params(params)
+            .map_err(|e| Error::InternalYubiKeyError(format!("{}", e)))?;
+        let csr = csr
+            .serialize_request_der()
+            .map_err(|e| Error::InternalYubiKeyError(format!("{}", e)))?;
 
         Ok(csr)
     }
 
     /// Provisions the YubiKey with a new certificate generated on the device.
     /// Only keys that are generated this way can use the attestation functionality.
-    pub fn provision(
+    pub fn provision<KT: yubikey_signer::KeyType>(
```

Comment:
I think we hide this away and just provide two non-generic functions for p256 and p384